### DISCUSSION TITLE
How to establish a WebTransport connection between a JavaScript Client and a ASP.NET Kestrel server?

### DISCUSSION THREAD

**Discussion by iCJee** [dotnet/aspnetcore#50985] [Q&A]
Hello,
 
I have tried running the Kestrel WebTransport server samples that are provided in the repository and connecting using JavaScript, 
I took the certificate hash string that's printed when running the server and ran the following on client:

let x = new WebTransport("https://localhost:5003/",{serverCertificateHashes:[{algorithm:"sha-256",value:Uint8Array.from(atob("GENEREATED_STRING_HASH_HERE"),c=>c.charCodeAt(0))}]})

also tried adding ignore the generated certificate hash and origin force on QUIC flags when running chromium but ended up receiving the either a CONNECTION_RESET error or QUIC_IETF_GQUIC_ERROR_MISSING.

Can anyone clarify what am I missing here and what's the correct way to do so?

P.S, I've been following the blog discussing WebTransport on Kestrel here: 
https://devblogs.microsoft.com/dotnet/experimental-webtransport-over-http-3-support-in-kestrel/

And there seems to be radio silence on the matter for the recent year, are there any upcoming updates to be expected?

Thanks!

**Comment by BrennanConroy** [comment]
The WebTransport spec has changed since we last worked on it and the server implementation likely doesn't work anymore. For example, I don't think the draft header we're looking for will match with the most recent versions of the spec.
https://github.com/dotnet/aspnetcore/blob/85ae25913fe4c26429abe21c5919c165c79707c3/src/Servers/Kestrel/Core/src/Internal/WebTransport/WebTransportSession.cs#L41
You could try fixing just that header check locally and seeing if it works, or let us know if there is more work needed.

**Comment by iCJee** [comment]
> The WebTransport spec has changed since we last worked on it and the server implementation likely doesn't work anymore. For example, I don't think the draft header we're looking for will match with the most recent versions of the spec.
> 
> https://github.com/dotnet/aspnetcore/blob/85ae25913fe4c26429abe21c5919c165c79707c3/src/Servers/Kestrel/Core/src/Internal/WebTransport/WebTransportSession.cs#L41
> 
> 
> You could try fixing just that header check locally and seeing if it works, or let us know if there is more work needed.

I tried to check it locally and it doesn't seem to work, probably does require more work,
is there an existing example of a working implementation of WebTransport (Doesn't have to be Kestrel or .NET) anyone is aware of?

P.S Is there a date or ETA on when the spec will cease to become a draft and will become solid?

**Comment by MeydanOzeri** [comment]
any update ?

**Comment by AliKhalili** [comment]
Recently, while experimenting with the `WebTransportInteractiveSampleApp`, I encountered the same issue on both Windows 10 and macOS. I noticed that only Windows 11 and Linux support HTTP/3 and QUIC(see [here](https://learn.microsoft.com/en-us/aspnet/core/fundamentals/servers/kestrel/http3?view=aspnetcore-8.0#http3-requirements) also), making it impossible to test WebTransport over HTTP/3 on those operating systems. Additionally, the current WebTransport implementation is at draft 2, whereas most clients such as Chrome and Firefox are still at version 2 or 3, indicating that it's not yet outdated.
